### PR TITLE
fs: add backtrace to where file opens

### PR DESCRIPTION
**Pull request by XuNeo**
## Summary

Add backtrace to where file is open'ed. It's useful to debug fd leak.

```bash
nsh> fdinfo

pid:0
FD  OFLAGS  TYPE POS       PATH
0   3       1    0         /dev/console 0x00000000004aa3e2 0x000000000048c793 0x000000000048ef7f 0x000000000048f23e 0x000000000045ff1f 0x000000000045ddde 0x000000000045d6d0 0x00007cd386229d90
1   3       1    0         /dev/console 0x00000000004aa3e2 0x000000000048dace 0x000000000048c44d 0x000000000048ca37 0x000000000048ca63 0x000000000045ff37 0x000000000045ddde 0x000000000045d6d0
2   3       1    0         /dev/console 0x00000000004aa3e2 0x000000000048dace 0x000000000048c44d 0x000000000048ca37 0x000000000048ca63 0x000000000045ff46 0x000000000045ddde 0x000000000045d6d0

pid:1
FD  OFLAGS  TYPE POS       PATH

pid:2
FD  OFLAGS  TYPE POS       PATH
0   3       1    0         /dev/console 0x00000000004aa3e2 0x000000000048dace 0x000000000048db25 0x000000000048c937 0x0000000000460050 0x0000000000461d9f 0x0000000000461adc 0x0000000000461c79
```
## Impact
No.
## Testing
sim:citest

### CODE REVIEW

**Review by acassis — CHANGES_REQUESTED** [review]
Please include Documentation/ about this new feature!

**Comment by g2gps** [comment]
This is a really nice feature. I'm wondering if it would be possible to add it as part of the output printed when an assertion occurs, in `misc/assert.c`?

File: include/nuttx/fs/fs.h [hunk 1]
```diff
@@ -475,6 +475,10 @@ struct file
 #ifdef CONFIG_FDCHECK
   uint8_t           f_tag_fdcheck; /* File owner fdcheck tag, init to 0 */
 #endif
+
+#if CONFIG_FS_BACKTRACE > 0
+  FAR void         *backtrace[CONFIG_FS_BACKTRACE]; /* Backtrace for caller */
```
Comment (anchao) [inline include/nuttx/fs/fs.h:480]:
```suggestion
  FAR void         *f_backtrace[CONFIG_FS_BACKTRACE]; /* Backtrace for caller */
```

File: fs/vfs/fs_dup2.c [hunk 2]
```diff
@@ -174,6 +174,7 @@ int file_dup3(FAR struct file *filep1, FAR struct file *filep2, int flags)
   filep2->f_tag_fdcheck = filep1->f_tag_fdcheck;
 #endif
 
+  FS_ADD_BACKTRACE(filep2->backtrace);
```
Comment (anchao) [inline fs/vfs/fs_dup2.c:177]:
```suggestion
  FS_ADD_BACKTRACE(filep2);
```

File: fs/vfs/Kconfig [hunk 3]
```diff
@@ -64,3 +64,16 @@ config SIGNAL_FD_NPOLLWAITERS
...
 endif # SIGNAL_FD
+
+config FS_BACKTRACE
+	int "VFS backtrace"
+	default 0
+	---help---
+		Add backtrace to vfs file open.
+
+config FS_BACKTRACE_SKIP
+	int "Backtrace to skip"
```
Comment (anchao) [inline fs/vfs/Kconfig:75]:
```suggestion
	int "Skip depth of backtrace"
```

File: fs/procfs/fs_procfsproc.c [hunk 4]
```diff
@@ -1297,13 +1303,28 @@ static ssize_t proc_groupfd(FAR struct proc_file_s *procfile,
           path[0] = '\0';
         }
 
+#if CONFIG_FS_BACKTRACE > 0
```
Comment (anchao) [inline fs/procfs/fs_procfsproc.c:1306]: move all backtrace logic into separate function and format backtrace after line 1300

File: fs/procfs/fs_procfsproc.c [hunk 5]
```diff
@@ -1283,7 +1282,14 @@ static ssize_t proc_groupfd(FAR struct proc_file_s *procfile,
 
   for (i = 0; i < count; i++)
     {
-      filep = files_fget(&group->tg_filelist, i);
+      FAR struct file *filep = files_fget(&group->tg_filelist, i);
```
Comment (anchao) [inline fs/procfs/fs_procfsproc.c:1285]: Keep variables defined at the beginning of the function

File: fs/procfs/fs_procfsproc.c (hunk 4 above)
Comment (anchao) [reply fs/procfs/fs_procfsproc.c:1306]: and add "BACKTRACE" label in column

```
FD  OFLAGS  TYPE POS       PATH
0   3       1    0         /dev/console 0x00000000004aa3e2
```

```
FD  OFLAGS  TYPE POS       PATH         BACKTRACE
0   3       1    0         /dev/console 0x00000000004aa3e2
```

**Comment by acassis** [comment]
ping @XuNeo

**Comment by XuNeo** [comment]
> Please include Documentation/ about this new feature!

Hi, any idea where should I add the documentation to?

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**